doc/developer/design/20210707_qgm_sql_high_level_representation.md

```diff
@@ -0,0 +1,726 @@
+
+#### Simple `GROUP BY`
+
+![Simple GROUP BY](qgm/simple-group-by.svg)
+
+
+#### `GROUP BY + HAVING`
+
+![GROUP BY + HAVING](qgm/simple-group-by-having.svg)
+
+Note that the having filter is just a regular predicate on the `Select` box ranging over the `Grouping` box.
+
+#### `GROUP BY` vs. `DISTINCT`
+
+As mentioned above, `GROUP BY` and `DISTINCT` use different representation in QGM. The following diagram
+shows a simple `DISTINCT` query.
+
+![DISTINCT](qgm/simple-distinct.svg)
+
+As shown below, the equivalent `GROUP BY` query is automatically converted into the query graph above
+during the normalization process.
+
+![GROUP BY DISTINCT before normalization](qgm/group-by-distinct-before.svg)
+
+![GROUP BY DISTINCT after normalization](qgm/group-by-distinct-after.svg)
+
+#### Comma join
+
+![Simple comma join](qgm/simple-comma-join.svg)
+
+#### Inner join
+
+![Simple inner join](qgm/simple-inner-join.svg)
+
+Note that the inner join above is semantically equivalent to the comma join in the previous example. Boxes 1 and 2
+represent the binary inner joins in the query, but they can be squashed into box 0, without altering the results of
+the query. In fact, the normalization step will simplify this query leaving it exactly as the one in the example above:
+
+![Simple inner join after normalization](qgm/simple-inner-join-after-normalization.svg)
+
+#### Outer join
+
+![Simple left join](qgm/simple-left-join.svg)
+
+![Simple right join](qgm/simple-right-join.svg)
+
+Note that in QGM there is no join direction, so left and right joins have the same exact representation. Only the type
+of the quantifiers change its order.
+
+Representing outer joins as explicit boxes goes against the goal of removing any sense of direction or join order
+from the query graph. For example, the following two queries are semantically equivalent but they lead to different
+query models where the join ordering is implied. We could add a normalization rule that always pushes inner joins
+through the preserving side of an outer join, assuming that it is always better to do inner joins first.
+
+![join](qgm/outer-inner-join.svg)
+![join](qgm/inner-outer-join.svg)
+
+However, we would still have the explicit join ordering issue if both joins in the query above where outer joins.
+To fix that issue, normalization will lift the preserving quantifier of an outer join to the parent `Select`
+making it a `Foreach` quantifier, leaving the `OuterJoin` box with just a `Foreach` quantifier, which
+will represent a special potentially-correlated operator that produces a `NULL` row if none of the rows
+produced by its input matches the predicate. With this normalization, the query mentioned above will look as
+follows:
+
+![join](qgm/normalized-outer-joins.svg)
+
+#### Cross join
+
+A `CROSS JOIN` can be represented as a `Select` box with no predicates as shown below:
+
+![Simple cross join](qgm/simple-cross-join.svg)
+
+#### CTEs
+
+![Simple CTE](qgm/simple-cte.svg)
```

Comment:
`Q1` is never used for name resolution purposes. When resolving the names of the symbols referenced in the ON-clause `Q2` and `Q3` are used as the base quantifiers for name resolution. Box 2 is only there to override the column names.

Consider the following example with the same query but without overriding the column names and one of references of the CTE:

![cte](https://user-images.githubusercontent.com/1282736/126756562-10353b36-b4f7-4b48-b450-d38a43827ead.png)

The box with the overridden column names has disappeared and the quantifier with the non-renamed use of A is now displaying `A` as its alias.

I understand why this is confusing. In order implementations some boxes may have a name, and quantifier aliases are only used for explicit alias in the query. So going back to the graph with the column alias, with that implementation, Box 2 would have `A` as its name, but not `Q1`. During name resolution, we would look at the input box name for quantifiers that don't have an explicit alias. With such implementation could report more meaningful error messages, since the name of the CTE is explicit in the CTE, such as:

```sql
with a(...) as (...) select * from a a1, a a2 where a1.f1 = a.f3;
error: table name `a` could not be resolved, `a1` and `a2` reference to a CTE with that name but the name is masked by the alias
```

With my proposal, the CTE names are kept in the `NameResolutionContext` and go away when that context goes out of scope. 

Perhaps, with this proposal I tried to simplify the code too much, by having a single place to keep table names, at the cost of making things a bit more confusing.

In any case, adding names in boxes is a trivial modification. Only `BaseTable` and `Select` may have names, the former will always have a name, the latter will only have a name when it represents a CTE. That can be modeled by a method `name` method in `QueryBox` that would return an optional reference to a name.